core/async/src/test_loop/test_loop_sender.rs

```diff
@@ -0,0 +1,155 @@
+        Self {
+            actor: self.actor.clone(),
+            pending_events_sender: self.pending_events_sender.clone(),
+            shutting_down: self.shutting_down.clone(),
+        }
+    }
+}
+
+impl<A> DelayedActionRunner<A> for TestLoopSender<A>
+where
+    A: 'static,
+{
+    fn run_later_boxed(
+        &mut self,
+        name: &str,
+        dur: Duration,
+        f: Box<dyn FnOnce(&mut A, &mut dyn DelayedActionRunner<A>) + Send + 'static>,
+    ) {
+        if self.shutting_down.load(Ordering::Relaxed) {
+            return;
+        }
+
+        let mut this = self.clone();
+        let callback = move || f(&mut this.clone().actor.lock().unwrap(), &mut this);
+        self.pending_events_sender.send_with_delay(
+            CallbackEvent {
+                description: format!("DelayedActionRunner {:?}", name),
+                callback: Box::new(callback),
+            },
+            dur,
+        );
+    }
+}
+
+impl<M, A> CanSend<M> for TestLoopSender<A>
+where
+    M: actix::Message + Send + 'static,
+    A: Actor + HandlerWithContext<M> + 'static,
+{
+    fn send(&self, msg: M) {
+        let this = self.clone();
+        let callback = move || {
+            this.actor.lock().unwrap().handle(msg, &mut this.clone());
+        };
+        self.pending_events_sender.send_with_delay(
+            CallbackEvent {
+                description: format!("Message {:?}", type_name::<M>()),
+                callback: Box::new(callback),
+            },
+            Duration::ZERO,
+        );
+    }
+}
+
+impl<M, R, A> CanSend<MessageWithCallback<M, R>> for TestLoopSender<A>
+where
+    M: actix::Message<Result = R> + Send + 'static,
+    A: Actor + HandlerWithContext<M> + 'static,
+    R: 'static,
+{
+    fn send(&self, msg: MessageWithCallback<M, R>) {
+        let this = self.clone();
+        let callback = move || {
+            let MessageWithCallback { message: msg, callback } = msg;
+            let result = this.actor.lock().unwrap().handle(msg, &mut this.clone());
+            callback(Ok(result));
+        };
+        self.pending_events_sender.send_with_delay(
+            CallbackEvent {
+                description: format!("Message {:?}", type_name::<M>()),
+                callback: Box::new(callback),
+            },
+            Duration::ZERO,
+        );
+    }
+}
+
+impl<A> TestLoopSender<A>
+where
+    A: Actor + 'static,
+{
+    pub fn new(
+        actor: A,
+        pending_events_sender: DelaySender<CallbackEvent>,
+        shutting_down: Arc<AtomicBool>,
+    ) -> Self {
+        Self { actor: Arc::new(Mutex::new(actor)), pending_events_sender, shutting_down }
+    }
+
+    pub fn actor(&self) -> MutexGuard<A> {
+        self.actor.lock().unwrap()
+    }
+
+    pub fn queue_start_actor_event(&mut self) {
+        let this = self.clone();
+        let callback = move || this.actor.lock().unwrap().start_actor(&mut this.clone());
+        self.pending_events_sender.send_with_delay(
```

Comment:
nit. can we add a method "send" to the sender where Duration::ZERO is implicit?

Reply:
Unfortunately this rabbit hole is quite deep. This ties in with the question for why do we need Mutex in single thread application.

The CanSend interface requires us to have the object (in our case `TestLoopSender<A>`) implement the Sync+Send interface. This isn't directly required for test loop but the Actix implementations of PeerManagerActor and PeerActor require this.

For `TestLoopSender<A>` to implement Sync+Send, rust requires actor A to implement Send which isn't always guaranteed. In our case since we know what's going on and we have higher control over the components, it's fine to tell rust "just assume A implements Send" or rather "just assume `TestLoopSender<A>` implements Sync+Send".

There's also the AsyncComputationSpawner etc. etc. for which the most generic interface requires Send.

I hope this explanation makes sense. To fry one's brain, I'd recommend to explore commenting that unsafe impl block code and seeing what breaks :)